Panic when opening a symlink

### Summary

I have a symlink to the `runtime/` in my `~/.config/helix` . I wanted to open a theme in `runtime/themes` inside Helix so I opened the file picker. The symlink showed up but not any files within the actual folder. When I put the pointer on the runtime symlink the preview box says `<File not found>`. When I press enter on the symlink Helix panics:
```
thread 'main' panicked at 'editor.open failed: Is a directory (os error 21)', helix-term/src/ui/mod.rs:185:18
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

This doesn't seem very mysterious. I guess following symlinks just isn't implemented yet in the file picker, Helix shouldn't crash because of it though.

### Reproduction Steps

<!-- Ideally provide a key sequence and/or asciinema.org recording. --> 

I tried this:

1. Open helix in the config folder.
2. Open the file picker and select the runtime symlink

I expected this to happen:

Helix would show the files in the folder the symlink points to.

Instead, this happened:

Helix panics.


### Helix log

<details><summary>~/.cache/helix/helix.log</summary>
These are the latest 10 lines of `helix.log`. Nothing looks relevant.

```
2022-04-24T11:49:36.554 helix_view::theme [WARN] Theme: invalid style attribute: modifers
2022-04-24T11:49:36.562 helix_view::editor [ERROR] Failed to initialize the LSP for `source.toml` { LSP not defined }
2022-04-24T11:50:09.881 helix_view::theme [WARN] Theme: invalid style attribute: modifers
2022-04-24T11:50:09.881 helix_view::theme [WARN] Theme: invalid style attribute: modifers
2022-04-24T11:50:09.905 helix_view::editor [ERROR] Failed to initialize the LSP for `source.toml` { LSP not defined }
2022-04-24T11:50:49.229 helix_view::theme [WARN] Theme: invalid style attribute: modifers
2022-04-24T11:50:49.230 helix_view::theme [WARN] Theme: invalid style attribute: modifers
2022-04-24T11:50:49.252 helix_view::editor [ERROR] Failed to initialize the LSP for `source.toml` { LSP not defined }
2022-04-24T11:58:26.687 helix_view::theme [WARN] Theme: invalid style attribute: modifers
2022-04-24T11:58:26.687 helix_view::theme [WARN] Theme: invalid style attribute: modifers

```

</details>


### Platform

Linux

### Terminal Emulator

alacritty

### Helix Version

22.03-127-g2c7f770a

--- a/helix-term/src/ui/mod.rs
+++ b/helix-term/src/ui/mod.rs
@@ -118,6 +118,7 @@ pub fn file_picker(root: PathBuf, config: &helix_view::editor::Config) -> FilePi
         .hidden(config.file_picker.hidden)
         .parents(config.file_picker.parents)
         .ignore(config.file_picker.ignore)
+        .follow_links(config.file_picker.follow_symlinks)
         .git_ignore(config.file_picker.git_ignore)
         .git_global(config.file_picker.git_global)
         .git_exclude(config.file_picker.git_exclude)
@@ -146,14 +147,13 @@ pub fn file_picker(root: PathBuf, config: &helix_view::editor::Config) -> FilePi
         let entry = entry.ok()?;
 
         // This is faster than entry.path().is_dir() since it uses cached fs::Metadata fetched by ignore/walkdir
-        let is_dir = entry.file_type().map(|ft| ft.is_dir()).unwrap_or(false);
-
+        let is_dir = entry.file_type().map_or(false, |ft| ft.is_dir());
         if is_dir {
             // Will give a false positive if metadata cannot be read (eg. permission error)
-            return None;
+            None
+        } else {
+            Some(entry.into_path())
         }
-
-        Some(entry.into_path())
     });
 
     // Cap the number of files if we aren't in a git project, preventing
@@ -175,9 +175,14 @@ pub fn file_picker(root: PathBuf, config: &helix_view::editor::Config) -> FilePi
             path.strip_prefix(&root).unwrap_or(path).to_string_lossy()
         },
         move |cx, path: &PathBuf, action| {
-            cx.editor
-                .open(path.into(), action)
-                .expect("editor.open failed");
+            if let Err(e) = cx.editor.open(path.into(), action) {
+                let err = if let Some(err) = e.source() {
+                    format!("{}", err)
+                } else {
+                    format!("unable to open \"{}\"", path.display())
+                };
+                cx.editor.set_error(err);
+            }
         },
         |_editor, path| Some((path.clone(), None)),
     )
@@ -281,8 +286,8 @@ pub mod completers {
             .collect()
     }
 
-    pub fn filename(_editor: &Editor, input: &str) -> Vec<Completion> {
-        filename_impl(input, |entry| {
+    pub fn filename(editor: &Editor, input: &str) -> Vec<Completion> {
+        filename_impl(editor, input, |entry| {
             let is_dir = entry.file_type().map_or(false, |entry| entry.is_dir());
 
             if is_dir {
@@ -314,8 +319,8 @@ pub mod completers {
             .collect()
     }
 
-    pub fn directory(_editor: &Editor, input: &str) -> Vec<Completion> {
-        filename_impl(input, |entry| {
+    pub fn directory(editor: &Editor, input: &str) -> Vec<Completion> {
+        filename_impl(editor, input, |entry| {
             let is_dir = entry.file_type().map_or(false, |entry| entry.is_dir());
 
             if is_dir {
@@ -338,7 +343,7 @@ pub mod completers {
     }
 
     // TODO: we could return an iter/lazy thing so it can fetch as many as it needs.
-    fn filename_impl<F>(input: &str, filter_fn: F) -> Vec<Completion>
+    fn filename_impl<F>(editor: &Editor, input: &str, filter_fn: F) -> Vec<Completion>
     where
         F: Fn(&ignore::DirEntry) -> FileMatch,
     {
@@ -370,6 +375,7 @@ pub mod completers {
 
         let mut files: Vec<_> = WalkBuilder::new(&dir)
             .hidden(false)
+            .follow_links(editor.config().file_picker.follow_symlinks)
             .max_depth(Some(1))
             .build()
             .filter_map(|file| {

--- a/helix-view/src/editor.rs
+++ b/helix-view/src/editor.rs
@@ -71,6 +71,9 @@ pub struct FilePickerConfig {
     /// Enables ignoring hidden files.
     /// Whether to hide hidden files in file picker and global search results. Defaults to true.
     pub hidden: bool,
+    /// Enables following symlinks.
+    /// Whether to follow symbolic links in file picker and file or directory completions. Defaults to true.
+    pub follow_symlinks: bool,
     /// Enables reading ignore files from parent directories. Defaults to true.
     pub parents: bool,
     /// Enables reading `.ignore` files.
@@ -94,6 +97,7 @@ impl Default for FilePickerConfig {
     fn default() -> Self {
         Self {
             hidden: true,
+            follow_symlinks: true,
             parents: true,
             ignore: true,
             git_ignore: true,

